### ISSUE TITLE
LT01: tsql: within custom data type a space is wrong introduced, code corruption

### ISSUE THREAD

**Issue by fmms** [sqlfluff/sqlfluff#4614]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have the following tsql
``` 
DECLARE @bks AS [UTIL].[ItemList]
```

### Expected Behaviour

it passes sqlfluff

### Observed Behaviour

spaces are introcued...
```
DECLARE @bks AS [UTIL] . [ItemList]
```

Afterwards, the code does not parse anymore...

### How to reproduce

```
PS > cat .\customDatatype.sql
DECLARE @bks AS [UTIL].[ItemList]
PS > sqlfluff fix --dialect tsql .\customDatatype.sql --force
==== finding fixable violations ====
== [customDatatype.sql] FAIL
L:   1 | P:  23 | LT01 | Expected single whitespace between quoted identifier and
                       | dot '.'. [layout.spacing]
L:   1 | P:  24 | LT01 | Expected single whitespace between dot '.' and data type
                       | identifier. [layout.spacing]
==== fixing violations ====
2 fixable linting violations found
FORCE MODE: Attempting fixes...
Persisting Changes...
== [customDatatype.sql] PASS
Done. Please check your files to confirm.
PS > cat .\customDatatype.sql
DECLARE @bks AS [UTIL] . [ItemList]
```

### Dialect

tsql

### Version

```
==== sqlfluff ====
sqlfluff:                2.0.2 python:                 3.11.0
implementation:        cpython verbosity:                   1
dialect:                  tsql templater:                 raw
rules:                                all
```

### Configuration

none

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by PoisnFang** [comment]
also experiencing this with postgres, any work arounds?

**Comment by bvanelli** [comment]
Adding a minimum reproducable script also in postgres here. Here is the script:

```sql
CREATE TYPE definitions.interval_type AS ENUM ('HOURLY', 'DAILY', 'WEEKLY');

CREATE OR REPLACE FUNCTION foo() RETURNS TABLE (
    interval_type definitions.interval_type
) LANGUAGE plpgsql AS
$BODY$
BEGIN
    RETURN QUERY SELECT 'DAILY'::definitions.interval_type;
END;
$BODY$;

SELECT interval_type FROM foo();

```

Running `sqlfluff lint --dialect postgres test.sql` on the file results in:

```
== [test.sql] FAIL                                                                                                                         
L:   4 | P:  30 | LT01 | Expected single whitespace between naked identifier and
                       | dot '.'. [layout.spacing]
L:   4 | P:  31 | LT01 | Expected single whitespace between dot '.' and data type
                       | identifier. [layout.spacing]
All Finished 📜 🎉!
```

Now let's try to fix the file with the default rules `sqlfluff fix --dialect postgres test.sql`:

```
==== finding fixable violations ====
== [test.sql] FAIL                                                                                                                         
L:   4 | P:  30 | LT01 | Expected single whitespace between naked identifier and
                       | dot '.'. [layout.spacing]
L:   4 | P:  31 | LT01 | Expected single whitespace between dot '.' and data type
                       | identifier. [layout.spacing]
==== fixing violations ====
2 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] ...
Attempting fixes...
Persisting Changes...
== [test.sql] FIXED
Done. Please check your files to confirm.
All Finished 📜 🎉!
```

Now the file cannot be parsed anymore with `sqlfluff lint --dialect postgres test.sql`:

```
== [test.sql] FAIL                                                                                                                         
L:   3 | P:   1 |  PRS | Line 3, Position 1: Found unparsable section: 'CREATE OR
                       | REPLACE FUNCTION foo() RETURNS...'
WARNING: Parsing errors found and dialect is set to 'postgres'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

A workaround is ignoring this rule on the file:

```diff
-    interval_type definitions.interval_type
+    interval_type definitions.interval_type -- noqa: LT01
```
Then sqllint goes through fine `sqlfluff lint --dialect postgres test.sql`:

```
All Finished 📜 🎉!
```

**Comment by peterbud** [comment]
With the latest `sqlfluff` this is working properly, and LT01 is not triggered.

Original:
```sql
DECLARE @bks AS [UTIL].[ItemList]
```

```bash
sqlfluff format test.sql -d tsql
==== finding fixable violations ====
== [test.sql] FAIL                                                                                                                                                                
L:   1 | P:  24 | CP05 | Datatypes must be consistently upper case.
                       | [capitalisation.types]
== [test.sql] FIXED
1 fixable linting violations found
```

Result:
```sql
DECLARE @bks AS [UTIL].[ITEMLIST]
```

**Comment by peterbud** [comment]
Could one of the maintainer close this issue pls? I don't have access to do that. @alanmcruickshank @WittierDinosaur ?
